go-controller/pkg/node/udn_isolation.go

```diff
@@ -293,52 +295,77 @@ func (m *UDNHostIsolationManager) updateKubeletCgroup() error {
+		return err
 	}
+
 	defer func() {
 		if err != nil {
-			conn.Close()
+			if err := conn.Close(); err != nil {
+				klog.Errorf("Error closing dbus connection for UDN isolation: %v", err)
+			}
 		}
 	}()
 
-	err = conn.Subscribe()
+	// Only use EXTERNAL method, and hardcode the uid (not username)
+	// to avoid a username lookup (which requires a dynamically linked
+	// libc)
+	methods := []dbus.Auth{dbus.AuthExternal(strconv.Itoa(os.Getuid()))}
+
+	err = conn.Auth(methods)
 	if err != nil {
-		return fmt.Errorf("failed to subscribe to systemd events: %w", err)
+		return err
 	}
-	// interval is important here as we need to catch the restart state, before it is running again
-	events, errChan := conn.SubscribeUnitsCustom(50*time.Millisecond, 0, func(u1, u2 *dbus.UnitStatus) bool { return *u1 != *u2 },
-		func(s string) bool {
-			return s != "kubelet.service"
-		})
+
+	signalChan := make(chan *dbus.Signal, 10)
+	conn.Signal(signalChan)
+
 	// run until context is cancelled
 	go func() {
-		waitingForActive := false
 		for {
 			select {
 			case <-ctx.Done():
-				conn.Close()
+				if err := conn.Close(); err != nil {
+					klog.Errorf("Error closing dbus connection for UDN isolation: %v", err)
+				}
 				return
-			case event := <-events:
-				for _, status := range event {
-					if status.ActiveState != "active" {
-						waitingForActive = true
-					} else if waitingForActive {
-						klog.Infof("Kubelet was restarted, re-applying UDN host isolation")
-						err = m.updateKubeletCgroup()
-						if err != nil {
-							klog.Errorf("Failed to re-apply UDN host isolation: %v", err)
-						} else {
-							waitingForActive = false
+			case signal := <-signalChan:
+				klog.V(5).Infof("D-Bus event received: %#v", signal)
+				// Extract unit name from path
+				unitPath := signal.Path
+				parts := strings.Split(string(unitPath), "/")
+				if len(parts) < 6 || parts[4] != "unit" {
+					continue
+				}
+				escapedUnit := parts[5]
+				unitName := strings.ReplaceAll(escapedUnit, "_2e", ".")
+
+				if unitName == "kubelet.service" {
+					changes := signal.Body[1].(map[string]dbus.Variant)
+					if state, exists := changes["ActiveState"]; exists {
+						newState := state.Value().(string)
+						if newState == "active" {
+							klog.Info("Kubelet restarted, re-applying isolation")
+							if err := m.updateKubeletCgroup(); err != nil {
+								klog.Errorf("Failed to re-apply isolation: %v", err)
+							}
 						}
 					}
 				}
-			case err := <-errChan:
-				klog.Errorf("Systemd listener error: %v", err)
 			}
 		}
 	}()
+
+	conn.BusObject().Call("org.freedesktop.DBus.AddMatch", 0,
+		"type='signal',interface='org.freedesktop.DBus.Properties',member='PropertiesChanged'")
+
+	sigObj := conn.Object("org.freedesktop.systemd1", dbus.ObjectPath("/org/freedesktop/systemd1"))
+	if err := sigObj.Call("org.freedesktop.systemd1.Manager.Subscribe", 0).Store(); err != nil {
+		return fmt.Errorf("failed to subscribe to systemd: %w", err)
+	}
```

Comment:
why is this part set up after the signal handler is added?

Reply:
so this is the same way it happens in the coreos code. There they call "dispatch()" function which sets up the signal, and then they Subscribe after. My understanding is you want to be able to receive and handle events before you start subscribing, so that you dont miss any events.

Reply:
can't say I understand hoe that helps not missing events and why it is important on startup, but also not a big deal :)